src/Database/Database.php

```diff
@@ -0,0 +1,739 @@
+    //     return \reset($results);
+    // }
+
+    // /**
+    //  * @param string $collection
+    //  * @param array $options
+    //  *
+    //  * @return Document
+    //  */
+    // public function findLast(string $collection, array $options)
+    // {
+    //     $results = $this->find($collection, $options);
+    //     return \end($results);
+    // }
+
+    // /**
+    //  * @param array $options
+    //  *
+    //  * @return int
+    //  */
+    // public function count(array $options)
+    // {
+    //     $options = \array_merge([
+    //         'filters' => [],
+    //     ], $options);
+
+    //     $results = $this->adapter->count($options);
+
+    //     return $results;
+    // }
+
+    // /**
+    //  * @param array $data
+    //  *
+    //  * @return Document|false
+    //  *
+    //  * @throws Exception
+    //  */
+    // public function overwriteDocument(array $data)
+    // {
+    //     if (!isset($data['$id'])) {
+    //         throw new Exception('Must define $id attribute');
+    //     }
+
+    //     $document = $this->getDocument($data['$collection'], $data['$id']); // TODO make sure user don\'t need read permission for write operations
+
+    //     $validator = new Authorization($document, 'write');
+
+    //     if (!$validator->isValid($document->getPermissions())) { // Check if user has write access to this document
+    //         throw new AuthorizationException($validator->getDescription()); // var_dump($validator->getDescription()); return false;
+    //     }
+
+    //     $new = new Document($data);
+
+    //     if (!$validator->isValid($new->getPermissions())) { // Check if user has write access to this document
+    //         throw new AuthorizationException($validator->getDescription()); // var_dump($validator->getDescription()); return false;
+    //     }
+
+    //     $new = $this->encode($new);
+
+    //     $validator = new Structure($this);
+
+    //     if (!$validator->isValid($new)) { // Make sure updated structure still apply collection rules (if any)
+    //         throw new StructureException($validator->getDescription()); // var_dump($validator->getDescription()); return false;
+    //     }
+
+    //     $new = new Document($this->adapter->updateDocument($this->getDocument(self::COLLECTIONS, $new->getCollection()), $new->getId(), $new->getArrayCopy()));
+
+    //     $new = $this->decode($new);
+
+    //     return $new;
+    // }
+
+    /**
+     * Add Attribute Filter
+     *
+     * @param string $name
+     * @param callable $encode
+     * @param callable $decode
+     *
+     * @return void
+     */
+    static public function addFilter(string $name, callable $encode, callable $decode): void
+    {
+        self::$filters[$name] = [
+            'encode' => $encode,
+            'decode' => $decode,
+        ];
+    }
+
+    public function encode(Document $collection, Document $document):Document
+    {
+        $rules = $collection->getAttribute('rules', []);
```

Comment:
Wondering if I've missed where the `rules` attribute is used. Should this be?:
```suggestion
        $rules = $collection->getAttribute('attributes', []);
```

Reply:
Nope, the encode/decode is still not done, these are leftovers from the old code.